source/contribute.rst

```diff
@@ -41,128 +41,146 @@ contributing back to a Doctrine project.
+branches should exist and are the possible choices:
+
+- ``2.3.x`` from which the next patch release will be tagged (2.3.2, 2.3.3, etc.);
+- ``2.4.x`` from which the next minor release will be tagged (2.4.0);
+- ``3.0.x`` from which the next major release will be tagged (3.0.0).
+
+To find out what's the latest version, you can check the "Releases"
+block on the project's Github page. Alternatively, you can check
+Packagist.
+
+Anything you contribute to a branch will be merged up to higher branches
+at some point. For example, if you contribute to ``2.3.x``, your changes
+will eventually land on ``2.4.x`` and ``3.0.x``.
+
+
+The general rules are as follows:
+
+- Patch releases should be as stable as possible. They fix bugs, but in
+  avoid making changes as far as possible.
+- Improvements and additions are being made on the upcoming minor
+  release branch.
+- An upgrade to a new minor version must not require any changes on the
+  user's side.
+- There must be a clear upgrade path from one major version to the next.
+  That means necessary changes on the user's side will be advertised
+  through deprecation notices, if anyway possible.
+- Deprecations may be added for new minor versions only.
+- Before a feature is removed (or an API changed) in a major version
+  release, a corresponding deprecation notice must be included in a
+  minor version release preceding it.
+- If a replacement feature and/or alternative API will be provided, it
+  must be available in the minor release that adds the deprecation for
+  the old feature.
+
+With that in mind:
+
+If you are touching something outside the source directory (usually
+``src``, sometimes ``lib``), and that thing is not ``composer.json``,
+you should probably target the lowest branch.
+There is no way that your contribution is going to create a bug for
+the end user, so let us get it on all 3 branches with merges up.
+Examples are: documentation, tests, build scripts, tooling configuration.
+
+If you are touching something inside the source directory, you may
+still contribute to the lowest branch if it is a bug fix, and if you
+are touching comments that are not phpdoc comments.
+
+If the comments are phpdoc comments… it depends. If the phpdoc was
+inaccurate, you should target the lowest branch. If the phpdoc was
+imprecise but technically correct, you should target the next minor
+branch.
+
+If you are doing a refactoring and you think you didn't change the
+behavior of the code, you should still target the next minor branch
+unless the refactoring is necessary for a bugfix. Let us not take any
+unnecessary risks.
+
+If you want to deprecate something, you should target the next minor
+branch: upgrading to a new patch version should not result in any
+additional deprecations.
+
+If you want to add a new feature, you should target the next minor
+branch, because it might have unintentional side effects that break existing code.
```

Comment:
The next minor version branch may include:
* Refactorings, unless they are necessary for bugfixes. This is to avoid unnecessary risks.
* Adding deprecations – remember to update the `UPGRADE.md` file as well!
* Adding new features and/or public APIs

Reply:
What about changing minimum dependency requirements?

Reply:
Changing minimum dependency requirements is OK on a patch branch if it's absolutely necessary to fix a bug. I don't think that's often the case, and I think we don't have to mention this here.